app/Http/Utilities/Email.php

```diff
@@ -2,17 +2,120 @@
+
+    /**
+     * Competition instance.
+     *
+     * @var \Gladiator\Models\Competition
+     */
+    protected $competition;
+
+    /**
+     * Northstar user
+     *
+     * @var object
+     */
+    protected $users;
+
+    /**
+     * Contest instance.
+     *
+     * @var \Gladiator\Models\Contest
+     */
+    public $contest;
+
+    /**
+     * Array of all processed messages for per user
+     *
+     * @var array
+     */
+    public $allMessages;
+
+    /**
+     * Constructor
+     */
+    public function __construct(Message $message, Contest $contest, Competition $competition, $users)
     {
-        //@TODO: I only tested this with competition, is there other things we want to replace?
-        $tokens = $model::$tokenizable;
-        foreach ($tokens as $token) {
-            $message->body = str_replace(':'.$token, $model->$token, $message->body);
-                //@TODO: handle the message subject.
-        }
+        $this->message = $message;
+        $this->contest = $contest;
+        $this->competition = $competition;
+        $this->users = $users;
+
+        $this->setupEmail();
+    }
+
+    /**
+     * Process all of the tokens for this email. New tokens should be defined here.
+     * @TODO - possibly move token stuff into it's own class.
+     *
+     * @param  object $user
+     * @return array $tokens
+     */
+    protected function defineTokens($user)
+    {
+        // @TODO - set defaults?
+        $tokens = [
+            ':competition_end_date' => $this->competition->competition_end_date,
+            ':leaderboard_msg_day' => $this->competition->leaderboard_msg_day,
+            ':first_name' => $user->first_name,
+        ];
+
+        return $tokens;
+    }
+
+    /**
+     * Process the message, replacing all tokens with their associated value.
+     *
+     * @param  array $tokens
+     * @param  \Gladiator\Models\Message $message
+     * @return \Gladiator\Models\Message $preparedMessage
+     */
+    protected function processMessage($tokens, $message)
+    {
+        $preparedMessage = clone $message;
+
+        // // @TODO - maybe loop through message properties and run the replace on each one.
+        $preparedMessage->body = $this->replaceTokens($tokens, $message->body);
+        $preparedMessage->subject = $this->replaceTokens($tokens, $message->subject);
+
+        return $preparedMessage;
+    }
 
-        return $message;
+    /**
+     * Handles the string replacement
+     *
+     * @param  array $tokens
+     * @param  string $string
+     * @return string
+     */
+    protected function replaceTokens($tokens, $string)
+    {
+        return str_replace(array_keys($tokens), array_values($tokens), $string);
+    }
+
+    /**
+     * Builds the email object.
+     */
+    protected function setupEmail()
+    {
+        // Each user gets it's own processed message
+        foreach ($this->users as $key => $user) {
+            $this->allMessages[$key]['user'] = $user;
```

Comment:
yea i noticed that but wanted to leave it open to the possibility...i mean if any message has user content in it, then this will need to happen (or we can like scan for the existence of a user token and only process it here if we need, but that seemed extra to me) . Or we can make the decision that user content can't be in the message body.